File: plugin/src/main/java/io/github/kdy05/physicalFighters/ability/Ckyomi.java
package io.github.kdy05.physicalFighters.ability;

import io.github.kdy05.physicalFighters.module.InvincibilityManager;
import org.bukkit.entity.LivingEntity;
import io.github.kdy05.physicalFighters.core.Ability;
import io.github.kdy05.physicalFighters.core.AbilitySpec;
import io.github.kdy05.physicalFighters.core.EventManager;
import io.github.kdy05.physicalFighters.util.EventData;

import org.bukkit.event.Event;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import io.github.kdy05.physicalFighters.util.PotionEffectFactory;

public class Ckyomi extends Ability {
    public Ckyomi() {
        super(AbilitySpec.builder("츠쿠요미", Type.Passive_AutoMatic, Rank.A)
                .guide("상대를 공격하면 상대에게 5초간 혼란 효과와 디버프를 줍니다.")
                .build());
        EventManager.onEntityDamageByEntity.add(new EventData(this));
    }

    @Override
    public int A_Condition(Event event, int CustomData) {
        EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
        if (!InvincibilityManager.isDamageGuard() && isOwner(event0.getDamager())
                && event0.getEntity() instanceof LivingEntity) {
            return 0;
        }
        return -1;
    }

    @Override
    public void A_Effect(Event event, int CustomData) {
        EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
        LivingEntity entity = (LivingEntity) event0.getEntity();
        entity.addPotionEffect(PotionEffectFactory.createNausea(100, 0));
        entity.addPotionEffect(PotionEffectFactory.createWeakness(100, 0));
        entity.addPotionEffect(PotionEffectFactory.createBlindness(100, 0));
    }
}

File: abilities/src/main/kotlin/io/github/kdy05/physicalFighters/ability/impl/Ckyomi.kt
package io.github.kdy05.physicalFighters.ability.impl

import io.github.kdy05.abilityAPI.ability.AbilityMeta
import io.github.kdy05.abilityAPI.rank.Rank
import io.github.kdy05.abilityAPI.skill.SkillContext
import io.github.kdy05.physicalFighters.ability.PFAbility
import io.github.kdy05.physicalFighters.util.PotionEffectFactory
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player

@AbilityMeta(
    name = "츠쿠요미",
    rank = Rank.A,
    guide = ["상대를 공격하면 상대에게 5초간 혼란 효과와 디버프를 줍니다."]
)
class Ckyomi(owner: Player, context: SkillContext) : PFAbility(owner, context) {
    override fun buildSkills() = listOf(object : PassiveSkill() {
        override fun register() {
            onEntityDamage { e ->
                val entity = e.entity as? LivingEntity ?: return@onEntityDamage
                entity.addPotionEffect(PotionEffectFactory.createNausea(100, 0))
                entity.addPotionEffect(PotionEffectFactory.createWeakness(100, 0))
                entity.addPotionEffect(PotionEffectFactory.createBlindness(100, 0))
            }
        }
    })
}
